File: rest-service/src/main/java/org/veupathdb/service/multiblast/model/ErrorMap.java
package org.veupathdb.service.multiblast.model;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import org.veupathdb.service.multiblast.model.blast.ToolOption;

public final class ErrorMap extends HashMap<String, List<String>>
{
  public void putError(ToolOption key, String error) {
    putError(key.toString(), error);
  }

  public void putError(String key, String error) {
    computeIfAbsent(key, this::newList).add(error);
  }

  public void putError(String prefix, String key, String error) {
    computeIfAbsent(prefix + "." + key, this::newList).add(error);
  }

  public void putError(String prefix, String key, List<String> errors) {
    computeIfAbsent(prefix + "." + key, this::newList).addAll(errors);
  }

  private List<String> newList(String ignored) {
    return new ArrayList<>();
  }
}

File: service-report/src/main/kotlin/org/veupathdb/service/mblast/report/valid/error-map.kt
package org.veupathdb.service.mblast.report.valid

typealias ErrorMap = MutableMap<String, MutableList<String>>

fun NewErrorMap(): ErrorMap = HashMap()

fun ErrorMap.appendError(path: String, error: String) {
  computeIfAbsent(path) { ArrayList(1) }
    .add(error)
}